Federate reports

mouse-reeve opened the pull request:
Reports should be sent (if the user wants) to the home instance for remote users. I'm using mastodon's Flag activity as a template: https://docs.joinmastodon.org/spec/activitypub/#Flag

Fixes #962 

Components:
 - [x] Fix the report model so that `user` means the user doing the reporting, not the reported user
 - [x] Make the report model activitypub-y
   - [x] Produce the easy parts of the activitypub object
   - [x] Get the "objects" part of the Flag activity in place
 - [x] Allow users to opt out of federating reports
 - [x] Broadcast reports
 - [x] Receive broadcasted reports 
 - [ ] Test between bookwyrm instances
 - [ ] Test between bookwyrm and mastodon instances

mouse-reeve committed:
Makes reports an activitypub model

mouse-reeve committed:
Changes field names in report model so the reporter is "user"

This is such an annoying change but it is objectively better. Just gotta
be real sure they didn't get mixed up anywhere along the way.

mouse-reeve committed:
Allow the user to decide if they want to send along a report

mouse-reeve committed:
Fixes some user/reported_user references

mouse-reeve committed:
UI for deciding to send reports to remote

Shows a checkbox for the user and lets the admin know

mouse-reeve committed:
Expands tests and fixes logic error in report recipients

mouse-reeve committed:
Merge branch 'main' into federate-reports

mouse-reeve committed:
Adds merge migration

mouse-reeve committed:
Fixes error in report tests

mouse-reeve committed:
Makes report button work without javascript

mouse-reeve committed:
Fixes static view of report form in no js mode

mouse-reeve committed:
Adds first-pass test for incoming reports

I'm writing this offline so I can't check what the actual Flag activity
spec looks like from Mastodon.

mouse-reeve committed:
Json view of admin report

mouse-reeve committed:
Merge branch 'main' into federate-reports

mouse-reeve committed:
Merge branch 'federate-reports' of github.com:bookwyrm-social/bookwyrm into federate-reports

mouse-reeve committed:
Adds merge migration

mouse-reeve committed:
Adds list of "object"s to activitypub flag object

mouse-reeve committed:
Get flagged status out of incoming Flag activity

mouse-reeve committed:
Converts status to many to many field on report model

This allows us to support mastodon reports that include multiple statuses

mouse-reeve committed:
Updates tests for reports

mouse-reeve committed:
Updates antispam automoderator to handle new statuses field

mouse-reeve committed:
Merge branch 'main' into federate-reports

hughrun commented on bookwyrm/models/report.py:
```diff
@@ -22,28 +24,61 @@
+    def object(self):
+        """Generate a list of reported objects in a format Mastodon will like"""
+        items = [self.reported_user.remote_id]
+        if self.statuses:
```
When I test this `self.statuses` is always `bookwyrm.Status.None` here. So the status is never sent to Mastodon. _However_ `statuses` obviously _isn't_ empty because in the BookWyrm admin interface I can see it. So I have no idea what is going on but it's weird and means statuses aren't passed through to federated servers with the report.

hughrun commented on bookwyrm/models/report.py:
```diff
@@ -22,28 +24,61 @@
+    )
+    statuses = models.ManyToManyField(
         "Status",
         null=True,
```
No need to set `null` on a `ManyToManyField`.

mouse-reeve committed:
Merge branch 'main' into federate-reports

mouse-reeve committed:
Remove null from report model

mouse-reeve committed:
Updates migrations

mouse-reeve committed:
Appeases ruff

mouse-reeve committed:
Avoid error when broadcasting

hughrun committed:
Merge branch 'main' into pull-3188-federate-reports

hughrun committed:
fixes

- fix incoming Mastodon reports janking everything up
- fix statuses not attaching to reports sent to external server

hughrun committed:
Merge branch 'main' into federate-reports

hughrun committed:
fix tests and tidy Flag activity

- remove irrelevant kwarg allowing external calls
- add mocks to tests

hughrun committed:
Merge branch 'federate-reports' into pull-3188-federate-reports

mouse-reeve committed:
Merge pull request #3937 from hughrun/pull-3188-federate-reports

Fix federate reports

mouse-reeve commented:
I tried reporting from Mastodon and got this error:
```
web-1                          | Internal Server Error: /user/mouse/inbox              
web-1                          | Traceback (most recent call last):                              
web-1                          |   File "/venv/lib/python3.11/site-packages/django/core/handlers/exception.py", line 55, in inner
web-1                          |     response = get_response(request)                 
web-1                          |                ^^^^^^^^^^^^^^^^^^^^^                  
web-1                          |   File "/venv/lib/python3.11/site-packages/django/core/handlers/base.py", line 197, in _get_response
web-1                          |     response = wrapped_callback(request, *callback_args, **callback_kwargs)
web-1                          |                ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
web-1                          |   File "/venv/lib/python3.11/site-packages/django/views/generic/base.py", line 105, in view
web-1                          |     return self.dispatch(request, *args, **kwargs)              
web-1                          |            ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^  
web-1                          |   File "/venv/lib/python3.11/site-packages/django/utils/decorators.py", line 48, in _wrapper
web-1                          |     return bound_method(*args, **kwargs)              
web-1                          |            ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^                        
web-1                          |   File "/venv/lib/python3.11/site-packages/django/views/decorators/csrf.py", line 65, in _view_wrapper
web-1                          |     return view_func(request, *args, **kwargs)       
web-1                          |            ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^        
web-1                          |   File "/venv/lib/python3.11/site-packages/django/utils/decorators.py", line 48, in _wrapper
web-1                          |     return bound_method(*args, **kwargs)            
web-1                          |            ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^              
web-1                          |   File "/app/bookwyrm/decorators.py", line 13, in wrap          
web-1                          |     return function(request, *args, **kwargs)   
web-1                          |            ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^       
web-1                          |   File "/venv/lib/python3.11/site-packages/django/views/generic/base.py", line 144, in dispatch
web-1                          |     return handler(request, *args, **kwargs)           
web-1                          |            ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^       
web-1                          |   File "/app/bookwyrm/views/inbox.py", line 72, in post         
web-1                          |     sometimes_async_activity_task(activity_json)    
web-1                          |   File "/app/bookwyrm/views/inbox.py", line 118, in sometimes_async_activity_task
web-1                          |     activity.action(allow_external_connections=False)           
web-1                          | TypeError: Flag.action() got an unexpected keyword argument 'allow_external_connections'
```

mouse-reeve committed:
Adds allow_external_connections to Flag aciton

mouse-reeve commented:
Confirmed that incoming reports are working from Mastodon! Woo!

mouse-reeve commented:
This works in both directions! But I see that I [added back a line you removed](https://github.com/bookwyrm-social/bookwyrm/pull/3188/changes/5563400fc38506207aa2aebf37439e6724059696#diff-9aae236670b715ef0dd44eeaf20fa1f6205bd0f132898721edb10aac80b52739R285) @hughrun -- is there a good reason you took that out that I'm missing?

hughrun commented:
I think i just got confused when I was tidying up and then maybe didn't double check. It's crucial that external connections are allowed.

mouse-reeve commented:
Gotcha! It's not used in the function so it may have tripped a linter up.

hughrun commented:
@mouse-reeve I remember now. I was thinking it's already `True` in `Verb.action` so there's no need to set it. But I had forgotten that when overriding the method the arguments still need to be specified because they'll be used by external calls. So without it, Inbox calls `Flag.action` with an arg that isn't defined. Since you added it back, that should now work - the failing test seems to be a missing merge migration.

mouse-reeve committed:
Merge branch 'main' into federate-reports

mouse-reeve committed:
Adds merge migration

mouse-reeve commented:
Unreal level of satisfying to finally merge a two year old PR. Thank you so much for getting it over the line
